test/upsert/mzcompose.py

```diff
@@ -102,3 +113,51 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
             ci_util.upload_junit_report(
                 "testdrive", Path(__file__).parent / junit_report
             )
+
+
+def workflow_failpoint(c: Composition) -> None:
+    """Test behaviour when upsert state errors"""
+
+    with c.override(
+        Clusterd(name="storage_cluster"),
+        Testdrive(no_reset=True, consistent_seed=True, default_timeout="10s"),
+    ):
+        print("Running failpoint workflow")
+
+        for failpoint in [
+            (
+                "fail_state_multi_put",
+                "failure while rehydrating state: Error putting values into state",
+            ),
+            (
+                "fail_state_multi_get",
+                "failure while fetching records from state: Error getting values from state",
+            ),
+        ]:
+            run_one_failpoint(c, failpoint[0], failpoint[1])
+
+
+def run_one_failpoint(c: Composition, failpoint: str, error_message: str) -> None:
+    failpoint_env = "FAILPOINTS={}=return".format(failpoint)
+    print(f">>> Running failpoint test for failpoint {failpoint_env}")
+
+    c.down(destroy_volumes=True)
+    dependencies = ["zookeeper", "kafka"]
+    c.up("storage_cluster", "materialized", *dependencies)
+    c.run("testdrive", "failpoint/01-setup.td")
+
+    with c.override(
+        # Start clusterd with failpoint
+        Clusterd(name="storage_cluster", environment_extra=[failpoint_env]),
```

Comment:
The dynamic `SET failpoints` in sql did not work, so spinning up separate clusterds with different environment variables for failpoint.